torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+            r"""
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
+
+            bucket_index = bucket.get_index()
+            rank = zero.global_rank
+            overlap_info = zero._overlap_info
+            if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+                overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+                params_per_rank = overlap_info.params_per_rank
+                params_per_bucket = overlap_info.params_per_bucket
+                if rank_to_update == rank:
+                    overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+                params_per_rank[rank_to_update].extend(bucket_params)
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
+
+            if rank_to_update == rank:
+                assert len(zero.optim.param_groups) == 1, \
+                    "Overlapping DDP with ZeRO only supports a single " \
+                    "parameter group"
+                # Construct the `gradients` input for the local optimizer step,
+                # which expects `None` in a list position to indicate that the
+                # corresponding parameter should not be updated
+                num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+                gradients: List[Optional[torch.Tensor]] = \
+                    [None for _ in range(num_local_optim_params)]
+                assert bucket_index in overlap_info.offsets, \
+                    f"Bucket index {bucket_index} was not assigned to rank " \
+                    f"{rank}"
+                offset = overlap_info.offsets[bucket_index]
+                bucket_gradients = bucket.get_per_parameter_tensors()
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+
+            device = bucket_params[0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
```

Comment:
does it make sense to hold on to the returned work handle here and in the last bucket wait for them all. So that when backward returns, all optimizer step comp/comm ops are inserted into the stream or correctly synchronized?

Reply:
Good catch. I think that this is needed.